src/Elastic.Apm.Azure.ServiceBus/MicrosoftAzureServiceBusDiagnosticListener.cs

```diff
@@ -0,0 +1,240 @@
+		private readonly Service _service;
+
+		public override string Name { get; } = "Microsoft.Azure.ServiceBus";
+
+		public MicrosoftAzureServiceBusDiagnosticListener(IApmAgent agent) : base(agent)
+		{
+			_realAgent = agent as ApmAgent;
+			_service = Service.GetDefaultService(agent.ConfigurationReader, agent.Logger);
+			_service.Framework = new Framework { Name = ServiceBus.SegmentName };
+		}
+
+		protected override void HandleOnNext(KeyValuePair<string, object> kv)
+		{
+			Logger.Trace()?.Log("Called with key: `{DiagnosticEventKey}'", kv.Key);
+
+			if (string.IsNullOrEmpty(kv.Key))
+			{
+				Logger.Trace()?.Log($"Key is {(kv.Key == null ? "null" : "an empty string")} - exiting");
+				return;
+			}
+
+			switch (kv.Key)
+			{
+				case "Microsoft.Azure.ServiceBus.Send.Start":
+					OnSendStart(kv, "SEND", _sendProperties);
+					break;
+				case "Microsoft.Azure.ServiceBus.Send.Stop":
+					OnStop(kv, _sendProperties);
+					break;
+				case "Microsoft.Azure.ServiceBus.Schedule.Start":
+					OnSendStart(kv, "SCHEDULE", _scheduleProperties);
+					break;
+				case "Microsoft.Azure.ServiceBus.Schedule.Stop":
+					OnStop(kv, _scheduleProperties);
+					break;
+				case "Microsoft.Azure.ServiceBus.Receive.Start":
+					OnReceiveStart(kv, "RECEIVE", _receiveProperties);
+					break;
+				case "Microsoft.Azure.ServiceBus.Receive.Stop":
+					OnStop(kv, _receiveProperties);
+					break;
+				case "Microsoft.Azure.ServiceBus.ReceiveDeferred.Start":
+					OnReceiveStart(kv, "RECEIVEDEFERRED", _receiveDeferredProperties);
+					break;
+				case "Microsoft.Azure.ServiceBus.ReceiveDeferred.Stop":
+					OnStop(kv, _receiveDeferredProperties);
+					break;
+				case "Microsoft.Azure.ServiceBus.Exception":
+					OnException(kv);
+					break;
+				default:
+					Logger.Trace()?.Log("`{DiagnosticEventKey}' key is not a traced diagnostic event", kv.Key);
+					break;
+			}
+		}
+
+		private void OnReceiveStart(KeyValuePair<string, object> kv, string action, PropertyFetcherCollection cachedProperties)
+		{
+			if (kv.Value is null)
+			{
+				Logger.Trace()?.Log("Value is null - exiting");
+				return;
+			}
+
+			var queueName = cachedProperties.Fetch(kv.Value,"Entity") as string;
+			if (MatchesIgnoreMessageQueues(queueName))
+				return;
+
+			var transactionName = queueName is null
+				? $"{ServiceBus.SegmentName} {action}"
+				: $"{ServiceBus.SegmentName} {action} from {queueName}";
+
+			var transaction = ApmAgent.Tracer.StartTransaction(transactionName, ServiceBus.Type);
+			transaction.Context.Service = _service;
+
+			// transaction creation will create an activity, so use this as the key.
+			var activityId = Activity.Current.Id;
+
+			if (!_processingSegments.TryAdd(activityId, transaction))
+			{
+				Logger.Trace()?.Log(
+					"Could not add {Action} transaction {TransactionId} for activity {ActivityId} to tracked segments",
+					action,
+					transaction.Id,
+					activityId);

```

Comment:
If this happens we should also make sure we don't leak the transaction. My best idea so far is to just call `End()` on it - better than leaking, but far from perfect. I just realize we have the same e.g. in `HttpDiagnosticListenerImplBase`. 

```suggestion
					activityId);
					transaction.End();
```

Reply:
Hmm, I'm not sure ending the transaction is the right thing to do, as it'll end up being sent to APM server and Elasticsearch with details that aren't a true reflection of the operation that it tried to instrument. 

This seems like an awkward issue to resolve for what is hopefully an edge case. Should we consider introducing something like an `Abandon()` method, that unsets it as the current transaction, and does't send to APM? Could this result in spans being sent to APM server though that are linked to this transaction? This seems unlikely because the service bus method that results in the diagnostic event is either synchronous or likely awaited.

My inclination is to leave this as is in the PR and to discuss more generally how we want to tackle this kind of issue. Sound good?

Reply:
> for what is hopefully an edge case.

Definitely an edge case - we have similar logging in other parts of the code, I have never seen those in any apps so far.

>  Should we consider introducing something like an `Abandon()` method

This sounds good to me, yes.

> My inclination is to leave this as is in the PR and to discuss more generally how we want to tackle this kind of issue. Sound good?

Agreed 👍